New Prisma SASE Pack and Integration

jdplant opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
related MR: [1575](https://code.pan.run/xsoar/content-test-conf/-/merge_requests/1575). 

## Description
Create a new Pack and Integration for Prisma SASE API

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

jdplant commented:
I fixed the formatting issue that was triggering the Flake8 failuer and I also updated the readme and secrets ignore file to clear up other build errors.

AradCarmi commented:
Hi @jdplant,
Thank you for contributing!
Apologies for the delay in reviewing your PR, the team will review your code as soon as possible.

Thanks again.

ShacharKidor commented:
Hi @jdplant, 
As discussed in our last meeting, please let me know when you are done adding the additional commands and the code is up to date and ready for review.

I'm available for every question here or via slack, 
Thank you :)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=20431) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=20431) before we can accept your contribution.<br/>**8** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: jdplant<br/>:white_check_mark: content-bot<br/>:white_check_mark: ilaner<br/>:white_check_mark: dorschw<br/>:white_check_mark: amshamah419<br/>:white_check_mark: merit-maita<br/>:white_check_mark: gal-forer<br/>:white_check_mark: yaakovpraisler<br/>:x: nkanon<br/>:x: MichaelYochpaz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=20431) it.</sub>

ShacharKidor reviewed (changes requested):
Hi @jdplant, 
Thank you for contributing this new pack, you did a great job!

I added my comments and suggestions below, please review them and correct the code in accordance. 

**A few more general comments:**
1. Please add a docstring for each function in the client - you can use the 'HelloWorld' integration as a reference for our docstring standards. 

2. Please add a docstring for each integration's command - you can use the 'HelloWorld' integration as a reference here as well.

3.  In all the commands you outputted the raw_response to both the context and the war room (as a readable output). I wonder if this is the right thing to do, it usually depends on the size and structure of the raw response. Let's talk about it offline - please reach out to me when you get to this point. 

4. In all the integration commands - please support the case of an empty raw response - add a suitable readable output for that case (you can find examples in different integration in the XSAOR marketplace). 

5. For each command - please add the context paths of the outputs to the YML file. By adding that users would be able to get those values in playbooks tasks. 
You can use the `demisto-sdk generate-outputs` command for that - for more information please see [generate_outputs](https://github.com/demisto/demisto-sdk/tree/master/demisto_sdk/commands/generate_outputs).

6. It would be great if you'll add command examples to the 'command_examples' file (please choose examples and arguments that return non-empty responses). 

7. After adding the outputs prefixes and the commands examples (5 + 6) please regenerate the readme file of the integration - you can do so by using the: `demisto-sdk generate-docs` command with the -e flag. For more information please see [generate_docs](https://github.com/demisto/demisto-sdk/tree/master/demisto_sdk/commands/generate_docs).

8. Please add a test playbook to test the API calls that are used in the integration. 
For more information about TPBs please see: [test-playbooks](https://xsoar.pan.dev/docs/integrations/test-playbooks).

I know that's a lot, but I'm here to help you - please don't hesitate to reach out to me with any questions. I'm available here or via slack. 

** I haven't had the chance to review your unit tests yet. I'll do it in the next few days. 

Thanks again.

jdplant commented:
> Hi @jdplant, Thank you for contributing this new pack, you did a great job!
> 
> I added my comments and suggestions below, please review them and correct the code in accordance.
> 
> **A few more general comments:**
> 
> 1. Please add a docstring for each function in the client - you can use the 'HelloWorld' integration as a reference for our docstring standards.
> 2. Please add a docstring for each integration's command - you can use the 'HelloWorld' integration as a reference here as well.
> 3. In all the commands you outputted the raw_response to both the context and the war room (as a readable output). I wonder if this is the right thing to do, it usually depends on the size and structure of the raw response. Let's talk about it offline - please reach out to me when you get to this point.
> 4. In all the integration commands - please support the case of an empty raw response - add a suitable readable output for that case (you can find examples in different integration in the XSAOR marketplace).
> 5. For each command - please add the context paths of the outputs to the YML file. By adding that users would be able to get those values in playbooks tasks.
>    You can use the `demisto-sdk generate-outputs` command for that - for more information please see [generate_outputs](https://github.com/demisto/demisto-sdk/tree/master/demisto_sdk/commands/generate_outputs).
> 6. It would be great if you'll add command examples to the 'command_examples' file (please choose examples and arguments that return non-empty responses).
> 7. After adding the outputs prefixes and the commands examples (5 + 6) please regenerate the readme file of the integration - you can do so by using the: `demisto-sdk generate-docs` command with the -e flag. For more information please see [generate_docs](https://github.com/demisto/demisto-sdk/tree/master/demisto_sdk/commands/generate_docs).
> 8. Please add a test playbook to test the API calls that are used in the integration.
>    For more information about TPBs please see: [test-playbooks](https://xsoar.pan.dev/docs/integrations/test-playbooks).
> 
> I know that's a lot, but I'm here to help you - please don't hesitate to reach out to me with any questions. I'm available here or via slack.
> 
> ** I haven't had the chance to review your unit tests yet. I'll do it in the next few days.
> 
> Thanks again.

Thanks Shachar!  Sorry for the delay in response.  I'll address these changes.

-jd

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,364 @@
+category: Network Security
```
hey @bakatzir I asked Shachar if we could consider marking this something other than Beta to identify that it has limited capability / early release?  We did previously talk about this but the more I think about it, the Beta tag may have our SP customers feeling like it isn't ready for their production use.  In this case, Beta just means that a full release with additional commands and such will be coming in the future.  Not a huge deal just wanted to see if we had any other options or naming we could use?

ShacharKidor replied:
Hi @bakatzir - would appreciate your opinion here.

bakatzir replied:
The SASE team has released the security policy mgmt APIs officially, as this integration supports only them currently, no need to roll it out as beta.

ShacharKidor replied:
Great! :) so I'm removing the beta key - thanks!

jdplant commented on Packs/PrismaSASE/README.md:
```diff
@@ -0,0 +1,21 @@
+---
+
+This content pack enables XSOAR to integrate with Palo Alto Networks Prisma Access. It includes 1 integrations.
+
```
Updated

jdplant commented on Packs/PrismaSASE/README.md:
```diff
@@ -0,0 +1,21 @@
+ - Get configuration jobs by ID
+
+
+The integration uses the Prisma SASE Multi-tenant API documented here: https://pan.dev
```
Made this recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/Pipfile:
```diff
@@ -0,0 +1,18 @@
+requests = "*"
+
+[requires]
+python_version = "3.7"
```
Made this recommended change.

jdplant commented on Packs/PrismaSASE/.secrets-ignore:
```diff
@@ -0,0 +1,2 @@
+http://base_url
+https://pan.dev
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/Pipfile:
```diff
@@ -0,0 +1,18 @@
+requests = "*"
+
+[requires]
+python_version = "3.7"
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+requests.packages.urllib3.disable_warnings()  # pylint: disable=no-member
+
+# CONSTANTS
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        for key in SECURITYRULE_FIELDS:
+            if key in keys:
+
+                if args.get(key) != "" and args.get(key) != "None":
+                    if key == 'profile_setting':
+                        if isinstance(args.get(key), list):
+                            val = args.get(key)
+                        else:
+                            val = [x.strip() for x in args.get(key).split(',')]   # type: ignore
+
+                        rule[key] = {'group': val}
+
+                    elif isinstance(SECURITYRULE_FIELDS.get(key), str):
+                        rule[key] = args.get(key)   # type: ignore
+
+                    elif isinstance(SECURITYRULE_FIELDS.get(key), list):
+                        if isinstance(args.get(key), list):
+                            val = args.get(key)
+                        else:
+                            val = [x.strip() for x in args.get(key).split(';')]   # type: ignore
```
Implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        return rule
+
+    def create_security_rule(self, rule: dict, folder: str, position: str, tsg_id: str):
+        uri = '/sse/config/v1/security-rules'
```
hey Shachar, this is true for the commands that I've added here.  However, the SASE API has several endpoints that would have a different URI prefix so I thought for future proofing to not make it a constant.  Maybe a good compromise is to have a constant for each of those prefixs and label them "CONFIG_URI" or something.  I'll do that now and let me know if you think there is a better way.

ShacharKidor replied:
OK, in that case, when we will add the additional commands we should add a constant for each relevant endpoint (if it will be used more than once in the code).

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    def create_security_rule(self, rule: dict, folder: str, position: str, tsg_id: str):
+        uri = '/sse/config/v1/security-rules'
+        access_token = self.get_access_token(tsg_id)
+
+        query_params = {
+            'folder': encode_string_results(folder),
+            'position': encode_string_results(position)
+        }
+
+        headers = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+            'Authorization': f"Bearer {access_token}"
+        }
```
Hey Shachar, the SASE API is multi-tenant capable and the tenant structure is hierarchical.  The TSG (tenant services group) is an identifier for a particular tenant.  A single API service account can have access to the root tenant and any number of sub-tenants underneath.  The scope / target of the API call is determined via the requested authorization token.   My intent was to enable someone to configure a single instance of the integration, but allow them to target any sub-tenant that the service account may have access to.  I really just followed what I saw in other token based authorizations but added the TSG variable in order to store multiple tokens.

jdplant replied:
I'm not sure I understand your suggestion on self._headers.

ShacharKidor replied:
Hi - thanks for your detailed explanation :) 
About the multi-tenant, can you estimate how many tenants will be used in parallel for the average user?

jdplant replied:
For an SP, it could be hundreds (the platform limits mention 100 Tenants per immediate parent but can also be 10 levels deep with a max of 1000 listed for the hierarchy.  These limits are not hard limits though as far as I understand.)

For a normal enterprise or mid sized customer, probably most would be under 5.

ShacharKidor replied:
OK thank you :) 
So I think we will keep your implementation that allows running each command on a different tenant by generating a separate access token for each tenant and storing it in the integration context. 
 I checked if we have any size limit/performance limitations for what can be stored in the integration's context, and it seems there isn't a technical limit (it can cause performance issues if the object is big (MBs) but that is not the case).

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+        access_token = self.get_access_token(tsg_id)
+
+        headers = {
+            'Content-Type': 'application/json',
+            'Authorization': f"Bearer {access_token}"
```
nope!  updated.  thanks:)

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    def get_access_token(self, tsg_id: str):
+
+        # Get existin API access token for specific TSG from integration context.
```
implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        tsg_access_token = f'{tsg_id}.access_token'
+        tsg_expiry_time = f'{tsg_id}.expiry_time'
+
+        '''
+        If there is an existing access token, and it has not expired, set it as the access token for this request
+        Else request a new access token for the provided TSG and store it in the integration context and add the TSG ID
+        as a prefix
+        '''
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        as a prefix
+        '''
+
+        if previous_token.get(tsg_access_token) and previous_token.get(tsg_expiry_time) > date_to_timestamp(datetime.now()):
+            return previous_token.get(tsg_access_token)
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE_description.md:
```diff
@@ -0,0 +1,16 @@
+
+
+## Prisma SASE API
+[https://pan.dev/sase](https://pan.dev/sase)
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE_description.md:
```diff
@@ -0,0 +1,16 @@
+This integration was integrated and tested with version 1.0 of Prisma SASE API
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE_description.md:
```diff
@@ -0,0 +1,16 @@
+## Configure Prisma SASE on Cortex XSOAR
+
+1. Navigate to **Common Services** > **Identity & Access** > **Servers & Services**.
+2. Add or create a service account with the proper privleges
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    oauth_url = demisto.params().get('oauth_url')
+    default_tsg_id = demisto.params().get('tsg_id')
+
+    LOG(f'Command being called is {demisto.command()}')
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        'prisma-access-list-address-objects': list_address_objects_command
+    }
+
+    command = demisto.command()
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    """
+        PARSE AND VALIDATE INTEGRATION PARAMS
+    """
+    # get the service API url
+    base_url = demisto.params()['url'].strip('/')
+    verify_certificate = not demisto.params().get('insecure', False)
+    proxy = demisto.params().get('proxy', False)
+
+    client_id = demisto.params().get('credentials').get('identifier')
+    client_secret = demisto.params().get('credentials').get('password')
+    oauth_url = demisto.params().get('oauth_url')
+    default_tsg_id = demisto.params().get('tsg_id')
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    # Log exceptions
+    except Exception as e:
+        return_error(f'Failed to execute {demisto.command()} command. Error: {str(e)}')
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,364 @@
+  name: insecure
+  required: false
+  type: 8
+description: Integrate Prisma SASE API
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,364 @@
+  version: -1
+configuration:
+- defaultvalue: https://api.sase.paloaltonetworks.com
+  display: API Gateway URL (https://api.sase.paloaltonetworks.com)
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,364 @@
+  name: tsg_id
+  required: false
+  type: 0
+- defaultvalue: https://auth.apps.paloaltonetworks.com/am/oauth2/access_token
+  display: Prisma API OAUTH URL
+  name: oauth_url
+  required: false
+  type: 0
```
This is the URL that is used for obtaining the OAUTH2 access token.  I added an additionalinfo parameter with this description.  I made this a parameter with a default in the event that this URL changes in the future.  Maybe better left as a constant?

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,364 @@
+  name: proxy
+  required: false
+  type: 8
+- display: ""
+  name: tsg_id
+  required: false
+  type: 0
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,364 @@
+  required: false
+  type: 8
+description: Integrate Prisma SASE API
+display: Prisma SASE
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+            headers=headers
+        )
+
+    def update_security_rule(self, rule: dict, folder: str, position: str, ruleid: str, tsg_id: str):
```
hoping to talk with you on this one.  Happy to make the change but I don't really understand.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    def get_access_token(self, tsg_id: str):
+
+        # Get existin API access token for specific TSG from integration context.
+        previous_token = get_integration_context()
+        tsg_access_token = f'{tsg_id}.access_token'
+        tsg_expiry_time = f'{tsg_id}.expiry_time'
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+                    raise DemistoException('Failed to parse json object from response: {}'.format(res.content),
+                                           exception)
+
+                if res.get('access_token'):
+                    expiry_time = date_to_timestamp(datetime.now(), date_format='%Y-%m-%dT%H:%M:%S')
+                    expiry_time += res.get('expires_in', 0) * 1000 - 10
+                    new_token = {
+                        tsg_access_token: res.get('access_token'),
+                        tsg_expiry_time: expiry_time
+                    }
+                    # store received token and expiration time in integration context
+                    set_integration_context(new_token)
+                    return res.get('access_token')
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+                if res.get('access_token'):
+                    expiry_time = date_to_timestamp(datetime.now(), date_format='%Y-%m-%dT%H:%M:%S')
+                    expiry_time += res.get('expires_in', 0) * 1000 - 10
```
Re-reading this code it doesn't make sense to have a default of 0.  If there is an access token response it shall include the expires_in field so no reason to ever default it to 0.  I used code from another integration, but I think this default isn't needed so I have removed it.

jdplant replied:
No and good catch.  I just copied the code.  I removed the multiplier.  I believe the value in the response is in seconds.

ShacharKidor replied:
@jdplant  - Ok great! 
I think I'll add back the 0 default value, cause without it we might get an exception there (when we will try to do None - 10). 
The meaning is that if the 'expires_in' key doesn't exist in the response, the token will be invalid immediately. What do you think?

jdplant replied:
ya that makes sense!

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+                    raise DemistoException('Failed to parse json object from response: {}'.format(res.content),
+                                           exception)
+
+                if res.get('access_token'):
```
I may need some help on this one.  Apologies I'm not clear on the best way to update.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+            except Exception as e:
+                raise DemistoException(f'Error occurred while creating an access token. Please check the instance configuration.'
+                                       f'\n\n{e.args[0]}')
```
Again, leftover code from different integration.  I've updated to include the whole e var.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+                                       f'\n\n{e.args[0]}')
+
+
+def test_module(client: Client, args: dict, default_tsg_id: str):
```
removed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    rule = client.build_security_rule(args)
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
ah yes you are right!  much cleaner and I've updated.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    else:
+        tsg_id = default_tsg_id
+
+    raw_response = client.create_security_rule(rule, args.get('folder'), args.get('position'), tsg_id)  # type: ignore
```
yes.  These args are sent to build_security_rule where they are parsed and formatted to be used in the create security rule API call.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    outputs = raw_response
+
+    return CommandResults(
+        outputs_prefix='PrismaAccess.CreatedSecurityRule',
```
So something like  outputs_prefix=f'{PA_OUTPUT_PREFIX}ConfigJob', ?

I think thats what you mean so I made that change.  If you are thinking something different please let me know I can update.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        "ip_netmask": args.get('ip_netmask'),
+        "name": args.get('name')}
+
+    if args.get('description'):
+        address_object["description"] = args.get('description')
+
+    if args.get('tag'):
+        address_object["tag"] = args.get('tag')
+
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    if args.get('tag'):
+        address_object["tag"] = args.get('tag')
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    # Create new Prisma Access address object
+
+    address_object = {
+        "ip_netmask": args.get('ip_netmask'),
```
yes thank you updated.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    )
+
+
+def edit_address_object_command(client: Client, args: dict, default_tsg_id: str):
```
same.  ip_netmask made manadatory.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    # Delete Prisma Access address object
+
+    if args.get('tsg_id'):
```
Implemented the recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    # Get address objects for a given Prisma Access Folder / Position
+
+    query_params = {
```
true.  Updated to be more descriptive.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        'folder': encode_string_results(args.get('folder'))
+    }
+
+    if args.get('tsg_id'):
```
implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    else:
+        tsg_id = default_tsg_id
+
+    if args.get('name'):
+        query_params["name"] = encode_string_results(args.get('name'))
```
implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    if args.get('name'):
+        query_params["name"] = encode_string_results(args.get('name'))
+
+    if args.get('limit'):
+        query_params["limit"] = encode_string_results(args.get('limit'))
+
+    if args.get('offset'):
+        query_params["offset"] = encode_string_results(args.get('offset'))
```
implemented recommended changes.  Yes offset is a number and I added a default of 0.

jdplant replied:
fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    raw_response = client.list_address_objects(query_params, tsg_id)  # type: ignore
+
+    outputs = (raw_response)['data']
```
implemented recommended changes.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    # Delete the specified security rule within the targeted Prisma Access tenant / TSG
+
+    rule_id = args.get('rule_id')
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
implemented recommended changes.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    # send query to the Prisma SASE Multi-tenant cloud management portal aggregate monitoring API
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
implemented recommended changes.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    rule = client.build_security_rule(args)
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
implemented recommended changes.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    # Get all security rules for a given Prisma Access Folder / Position
+
+    query_params = {
+        'folder': encode_string_results(args.get('folder')),
```
corrected to be a description.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    devices = [x.strip() for x in args.get('devices').split(',')]  # type: ignore
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    else:
+        tsg_id = default_tsg_id
+
+    raw_response = client.update_security_rule(rule, args.get(
```
Actually I think they aren't even applicable to this command so I have removed the parameters from the API call.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    # Trigger a configuration push for the identified Folder/Devices
+
+    devices = [x.strip() for x in args.get('devices').split(',')]  # type: ignore
```
Yes you are correct.  In fact, I have updated the name to folders to keep it more consistent and made it mandatory.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        'position': encode_string_results(args.get('position'))
+    }
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
Implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    else:
+        tsg_id = default_tsg_id
+
+    if args.get('name'):
+        query_params["name"] = encode_string_results(args.get('name'))
+
+    if args.get('limit'):
+        query_params["limit"] = encode_string_results(args.get('limit'))
+
+    if args.get('offset'):
+        query_params["offset"] = encode_string_results(args.get('offset'))
```
Implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    raw_response = client.list_security_rules(query_params, tsg_id)  # type: ignore
+
+    outputs = (raw_response)['data']
```
Implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    # Get single security rule using the rule name as a filter
+
+    query_params = {
+        'folder': encode_string_results(args.get('folder')),
```
since these are query parameters I thought to use this to ensure the param is URI encoded, since the folder can include spaces.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        "offset": 0
+    }
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
Implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    raw_response = client.list_security_rules(query_params, tsg_id)  # type: ignore
+
+    outputs = (raw_response)['data']
```
Implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    # List config jobs with a given id(s) for a given prisma access tenant
+
+    job_ids = args.get('id')
+    job_ids = job_ids.split(',')  # type: ignore
```
updated.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    job_ids = args.get('id')
+    job_ids = job_ids.split(',')  # type: ignore
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
implemented requested change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    for job_id in job_ids:
+
+        raw_response.append(client.get_config_jobs_by_id(tsg_id, job_id)['data'][0])  # type: ignore
```
implemented requested change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    # List all config jobs for a given prisma access tenant
+
+    if args.get('tsg_id'):
+        tsg_id = args.get('tsg_id')
+    else:
+        tsg_id = default_tsg_id
```
implemented requested change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        query_params["limit"] = encode_string_results(args.get('limit'))
+
+    if args.get('offset'):
+        query_params["offset"] = encode_string_results(args.get('offset'))
```
that makes sense.  Updated accordingly.

jdplant replied:
mistake.  fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    query_params = {}
+
+    if args.get('limit'):
+        query_params["limit"] = encode_string_results(args.get('limit'))
```
implemented recommended change.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+
+    raw_response = client.list_config_jobs(tsg_id, query_params)  # type: ignore
+
+    outputs = (raw_response)['data']
```
implemented recommended change.

jdplant commented:
Hey @ShacharKidor I believe this should be ready for another review.  

WIth respect to the overall feedback items you shared.  I covered all of those but has some comments about two of them:

In all the commands you outputted the raw_response to both the context and the war room (as a readable output). I wonder if this is the right thing to do, it usually depends on the size and structure of the raw response. Let's talk about it offline - please reach out to me when you get to this point.

JD: I updated this where it made sense, for many of them the full output makes sense as it is a single record.

In all the integration commands - please support the case of an empty raw response - add a suitable readable output for that case (you can find examples in different integration in the XSAOR marketplace).

JD: I did some testing here and found that the API will not return empty responses and will always return HTTP 400, so even if I wrote logic to handle empty response I don't think it will ever get used.

ShacharKidor commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,779 @@
+  name: proxy
+  required: false
+  type: 8
+- display: Default Tenant Services Group ID to use for API calls.
+  name: tsg_id
+  required: false
+  type: 0
```
Maybe better this way - what do you think? 
```suggestion
- display: Tenant Services Group ID
  name: tsg_id
  required: false
  type: 0
  additionalinfo: Default Tenant Services Group ID to use for API calls.
```

jdplant replied:
updated.

ShacharKidor reviewed (changes requested):
Hi @jdplant, 
I reviewed your changes - you did  **Great work!**

1. I added a few more comments\suggestions - please review them (below) and correct the code in accordance. 

2. Please go over the 'unresolved' comments from the last review - I reacted there.

3. I asked 2 questions on slack - let's discuss them there.

4. About the TestPlaybook - I can see that you included there only 6 out of 13 commands of the integration - what is the reason for that? 
Please also use the 'auto align' button to align it, download it again and run the `demisto-sdk format` command on it - should fix some of the errors in the build. 

5. Please handle the build's errors - let me know if you need help with that. 
About the secrets validation - please fix that ASAP (and in general please make sure again that all the urls\file hashes\ips\ etc that appear in the readmes \ tests are dummies. 

5. I may ask you to change a few things related to naming conventions and outputs prefixes - I'll consult with our TPMs and let you know soon. In addition, haven't had the chance to review your unit test yet - will update you about them too. 

I added @jrosenpanw as an assigned so she'll assign a member from her team to a tech docs review. 

I'm available on slack for every question :)

ShirleyDenkberg commented:
@jdplant All arguments need a description.  Let me know when the yml file has been updated with descriptions and I will review it.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+        Returns:
+            Outputs.
+        """
+        uri = f'/sse/config/v1/security-rules/{ruleid}'
```
Fixed.  Thanks!

ShacharKidor replied:
I replaced all the remain "/sse/config/v1/" strings with the CONFIG_URI_PREFIX const - in this [commit](https://github.com/demisto/content/pull/20431/commits/12fa148ec95b651b59ee1dd1e2b5f54d2b84ce71).

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+        Else request a new access token for the provided TSG and store it in the integration context and add the TSG ID
+        as a prefix
+        """
+        previous_token = get_integration_context()
```
its used in line 398 and 399.  Maybe I misunderstand?

jdplant replied:
actually the variable is used a few lines later but this definition looks like a mistake.  removed this line.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+        previous_token = integration_context.get(tsg_access_token)
+        previous_token_expiry_time = integration_context.get(tsg_expiry_time)
+
+        '''
+        If there is an existing access token, and it has not expired, set it as the access token for this request
+        Else request a new access token for the provided TSG and store it in the integration context and add the TSG ID
+        as a prefix
+        '''
```
fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+    Returns:
+        Outputs.
+    """
+    job_ids = args.get('id')
+    job_ids = job_ids.split(',')  # type: ignore
```
Handy function!  It let me reduce some code:)

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+    if limit := arg_to_number(args.get('limit', 10)):
+        query_params["limit"] = limit
+    if offset := arg_to_number(args.get('offset', 0)):
+        query_params["offset"] = encode_string_results(offset)
```
true!  Fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,779 @@
+      required: true
+    - description: Tenant services group ID
+      name: tsg_id
+    description: Delete security rule
```
fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,779 @@
+      required: true
+    - description: Tenant services group ID
+      name: tsg_id
+    description: Get specific config job by the jobid
```
fixed

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.yml:
```diff
@@ -0,0 +1,779 @@
+      name: negate_source
+    - description: Negate destination
+      name: negate_destination
+    description: Edit or edit existing security rule
```
Well you can either edit it, or you can edit it:)  Fixed, thanks:)

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+                        if isinstance(field_value, list):
+                            val = field_value
+                        else:
+                            val = [x.strip() for x in field_value.split(';')]   # type: ignore
```
agreed!  updated.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+                        if isinstance(field_value, list):
+                            val = field_value
+                        else:
+                            val = [x.strip() for x in field_value.split(',')]   # type: ignore
```
agreed!  fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+        outputs_prefix=f'{PA_OUTPUT_PREFIX}FoundSecurityRule',
+        outputs_key_field='id',
+        outputs=outputs,
+        readable_output=tableToMarkdown('Security Rules', outputs),
```
Yes this look better.  Updated.

ShacharKidor replied:
@jdplant  - Great :) I can see that you didn't add this in 2 commands (get_security_rule_by_name_command, query_agg_monitor_api_command) was it on purpose?

jdplant replied:
just missed.  Fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+    """
+    # get the service API url
+    params = demisto.params()
+    base_url = params.get('url').strip('/')
```
In the event a user adds the last slash when entering the URL.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+    params = demisto.params()
+    base_url = params.get('url').strip('/')
+    verify_certificate = not argToBoolean(params.get('insecure', False))
+    proxy = argToBoolean(params.get('proxy', False))
```
added.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+                                           exception)
+
+                if access_token := res.get('access_token'):
+                    expiry_time = date_to_timestamp(datetime.now(), date_format='%Y-%m-%dT%H:%M:%S')
```
no reason.  fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+        outputs_prefix='PrismaAccess.CreatedSecurityRule',
+        outputs_key_field='id',
+        outputs=outputs,
+        readable_output=tableToMarkdown('Security Rule Created', outputs),
```
I updated the outputs where needed (outputs that can have multiple rows).  Others I left this way as they will only include a single record.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,819 @@
+    else:
+        tsg_id = default_tsg_id
+
+    if args.get('query_data'):
+        query = json.loads(args.get('query_data'))  # type: ignore
+    else:
+        query = None
```
fixed.

jdplant commented on Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py:
```diff
@@ -0,0 +1,929 @@
+        for key in SECURITYRULE_FIELDS:
+            if key in keys:
+                field_value = args.get(key)
+                if field_value != "" and field_value != "None":
```
agreed. fixed.

jdplant commented:
> Hi @jdplant, I reviewed your changes - you did **Great work!**
> 
> 1. I added a few more comments\suggestions - please review them (below) and correct the code in accordance.

Done.  I think I got them all.

> 2. Please go over the 'unresolved' comments from the last review - I reacted there.

Should be good now, I think I got them.

> 3. I asked 2 questions on slack - let's discuss them there.

Done.

> 4. About the TestPlaybook - I can see that you included there only 6 out of 13 commands of the integration - what is the reason for that?

I'll double check.  I used the playbook generator but I guess I missed some.  Will fix and let you know when done.

>    Please also use the 'auto align' button to align it, download it again and run the `demisto-sdk format` command on it - should fix some of the errors in the build.

Will do.

> 5. Please handle the build's errors - let me know if you need help with that.

Will work on these and let you know once the commit is ready for another review.

>    About the secrets validation - please fix that ASAP (and in general please make sure again that all the urls\file hashes\ips\ etc that appear in the readmes \ tests are dummies.

Fixed.  I had fixed previously but lost my changes somehow.  They were not really exposures but adjusted anyway.

> 6. I may ask you to change a few things related to naming conventions and outputs prefixes - I'll consult with our TPMs and let you know soon. In addition, haven't had the chance to review your unit test yet - will update you about them too.

No worries.

> 
> I added @jrosenpanw as an assigned so she'll assign a member from her team to a tech docs review.
> 
> I'm available on slack for every question :)

ShirleyDenkberg commented:
@jdplant All the outputs also need descriptions.

jdplant commented:
> 

hey @ShirleyDenkberg I have added the output descriptions.  thanks.

ShirleyDenkberg commented:
@jdplant I reviewed and edited the integration yml file.  Please regenerate the integration readme with the updated text and the output descriptions you added.

jdplant commented:
> @jdplant I reviewed and edited the integration yml file. Please regenerate the integration readme with the updated text and the output descriptions you added.

hey @ShirleyDenkberg.  Done.  Thanks!

ShirleyDenkberg commented:
> > @jdplant I reviewed and edited the integration yml file. Please regenerate the integration readme with the updated text and the output descriptions you added.
> 
> hey @ShirleyDenkberg. Done. Thanks!

Hi.  The changes I made in the integration yml file are not reflected in the integration readme.

jdplant commented:
hey @ShacharKidor I had some issues with my dev setup and thought I did everything right to get it set back up.  However, after my last commit I am seeing this error in pre-commit checks "Error: Unable to locate a common ancestor between contrib/jdplant_jdplant_prisma_sase_pack and HEAD".  Honestly not sure where I went wrong.

ShirleyDenkberg commented:
@ShacharKidor @bakatzir @bmillerPaloAlto Doc review completed.

ShacharKidor commented:
@jdplant - As discussed yesterday, I'm merging the PR and a developer from the content team will take it from here and will add some improvements to the internal PR. 
We will keep you posted :) 

Thank you again for the amazing work!
It was a pleasure working with you. 

FYI - @Moishymat, @bakatzir.
